[discourse-gatekeeper] Migrate charm docs (#120)
canonical/kafka-k8s-operator: docs/how-to/h-upgrade.md

> **Note** This feature is available on Charmed Kafka K8s and Charmed ZooKeeper K8s from revisions 53 and 49, respectively. Upgrade from previous versions is **not supported**.

> Charm upgrades can include both upgrades to operator code (e.g. the revision used by the charm) and/or the workload version. Note that since the charm code pins a particular version of the workload, a charm upgrade may or may not involve also a workload version upgrade. In general, the following guide only applies for in-place upgrades that involve (at most) minor version upgrade of Kafka workload, e.g. between Kafka 3.4.x to 3.5.x. Major workload upgrades are generally **NOT SUPPORTED**, and they should be carried out using full cluster-to-cluster migrations. Please refer to the how-to guide about cluster migration [here](TODO) for more information on how this can be achieved.

Reviewer: Maybe it reads better with 

```
Charm upgrades can include both upgrades OF operator code (e.g. the revision used by the charm) and/or workload version.
```

Reviewer 2: ```suggestion
Charm upgrades can include both upgrades of operator code (e.g. the revision used by the charm) and/or the workload version. Note that since the charm code pins a particular version of the workload, a charm upgrade may or may not involve also a workload version upgrade. In general, the following guide only applies for in-place upgrades that involve (at most) minor version upgrade of Kafka workload, e.g. between Kafka 3.4.x to 3.5.x. Major workload upgrades are generally **NOT SUPPORTED**, and they should be carried out using full cluster-to-cluster migrations. Please refer to the how-to guide about cluster migration [here](TODO) for more information on how this can be achieved.
```

Reviewer: I would probably just skip "the" in "and/or the workload version". It feels to me that it reads better

> **Note** Even though the process will not stop anymore on each upgrade, 
> the charm will take care of coordinating the pod updates by restarting one unit at the time to not lose high-availability. 

> > This generally lets the Kafka service to continuously be up and running during an upgrade, although downtime should 

Reviewer: remove "to", or maybe better 

```suggestion
> This generally keeps Kafka service up and running during an upgrade, although downtime should 
```

### Failing upgrade

Before upgrading the unit, the charm will check whether the upgrade can be performed, e.g. this may mean:
1. Checking that the upgrade from the previous charm revision and Kafka version is allowed.
2. Checking that other external applications that Kafka depends on (e.g. ZooKeeper) are running the correct version.

Note that these checks are only possible after a refresh of the charm code, and therefore cannot be done upfront (e.g. during the `pre-upgrade-checks` action).
If some of these checks fail, the upgrade will be aborted. When this happens, the workload may still be operating (as only the operator may have failed) but we recommend to rollback the upgrade as soon as possible. 

> To roll back the upgrade, re-run steps two and three, using the revision taken in step one, i.e.:

Reviewer: given we call "Step 1" and "Step X", isn't it clearer to use the numbers? If we still want to use "one", I would update the heading to keep it consistent

Reviewer 2: I like using numbers in text as it seems easier to navigate and highlight important information. But we have to be consistent and follow our style guide - https://docs.ubuntu.com/styleguide/en#numbers. In this particular case, we can justify an exception for the Step numeration but it's a slippery slope.

Reviewer: To me this would definitely be included in 

> Exceptions to this rule include numbered lists and units of measurement.

I believe we should be consistent. If we really want to spell out the numbers, we should also do this in the [heading](https://github.com/canonical/kafka-k8s-operator/blob/f27818eccebafcd0188ce3d26c9195dc660e21c6/docs/how-to/h-upgrade.md?plain=1#L24) with "Step one: Collect". 

But really, to me those numbers look like a numbered list, IMHO

Reviewer 2: Let's do numbers for now.
I've added the `#` sign now. It looks good and will help to evade Vale linter warnings.

Reviewer: Nice! Sounds good!
